### PR TITLE
fix: fixes #26498 gather inplace update

### LINKED ISSUE #26498: gather

### PR DESCRIPTION
**Pull request by Kacper-W-Kozdon**
Replaces a rudimentary inplace update in gather with ivy.inplace_update.

## Related Issue

Closes #26498 

## Checklist

- [ ] ⏩ Did you add a function?
- [ ] ⏩ Did you add the tests?
- [ ] ✅ Did you run your tests and are your tests passing?
- [ ] ✅ Did pre-commit not fail on any check?
- [ ] ✅ Did you follow the steps we provided?

### CODE REVIEW

**Comment by Kacper-W-Kozdon** [comment]
## Reformatting Task Checklist
#### IMPORTANT NOTICE 🚨:
The [Ivy Docs](https://unify.ai/docs/ivy/) represent the ground truth for the task descriptions and this checklist should only be used as a supplementary item to aid with the review process.

#### LEGEND 🗺:
- ❌ :  Check item is not completed.
- ✅ :  Check item is ready for review.
- 🆘 :  Stuck/Doubting implementation (PR author should add comments explaining why).
- ⏩ :  Check is not applicable to function (skip).
- 🆗 :  Check item is already implemented and does not require any edits.

#### CHECKS 📑:
1. - [ ] ✅:  Remove all lambda and direct bindings for the backend functions in:
       - [ ] 🆗: [ivy/functional/backends/jax/general.py](https://github.com/unifyai/ivy/tree/main/ivy/functional/backends/jax/general.py).
       - [ ] 🆗: [ivy/functional/backends/numpy/general.py](https://github.com/unifyai/ivy/tree/main/ivy/functional/backends/numpy/general.py).
       - [ ] 🆗: [ivy/functional/backends/tensorflow/general.py](https://github.com/unifyai/ivy/tree/main/ivy/functional/backends/tensorflow/general.py).
       - [ ] ✅: [ivy/functional/backends/torch/general.py](https://github.com/unifyai/ivy/tree/main/ivy/functional/backends/torch/general.py).
2. - [ ] 🆗: Implement the following if they don't exist:
       1. - [ ]  🆗: The `ivy.Array` instance method in [ivy/data_classes/array/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/general.py).
       2. - [ ]  🆗: The `ivy.Array` special method in [ivy/data_classes/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py).
       3. - [ ]  ⏩: The `ivy.Array` reverse special method in [ivy/data_classes/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py).
       4. - [ ] 🆗: The `ivy.Container` static method in [ivy/data_classes/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py).
       5. - [ ] 🆗: The `ivy.Container` instance method in [ivy/data_classes/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py).
       6. - [ ]  🆗:  The `ivy.Container` special method in [ivy/data_classes/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py).
       7. - [ ]  ⏩: The `ivy.Container` reverse special method in [ivy/data_classes/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py).
3. - [ ] 🆗:  Make sure that the aforementioned methods are added into the correct category-specific parent class, such as  `ivy.ArrayWithElementwise`,  `ivy.ContainerWithManipulation`  etc.
4. - [ ] 🆗:  Correct all of the  [Function Arguments and the type hints](https://unify.ai/docs/ivy/overview/deep_dive/function_arguments.html#function-arguments) for every function  **and**  its  _relevant methods_, including those you did not implement yourself.
5. - [ ] 🆗: Add the correct  [Docstrings](https://unify.ai/docs/ivy/overview/deep_dive/docstrings.html#docstrings)  to every function  **and**  its  _relevant methods_, including those you did not implement yourself. The following should be added:
       1. - [ ] 🆗:   <a name="ref1"></a> The function's [Array API standard](https://data-apis.org/array-api/latest/index.html) description in [ivy/functional/general.py](https://github.com/unifyai/ivy/blob/main/ivy/functional/ivy/general.py). If the function is not part of the Array API standard then a description of similar style should be added to the same file.
	The following modifications should be made to the description:
              - [ ] 🆗:  Remove type definitions in the `Parameters` and `Returns` sections.
              - [ ] 🆗:  Add `out` to the `Parameters` section if function accepts an `out` argument.
              - [ ] 🆗:  Replace `out` with `ret` in the `Returns` section.
       2. - [ ] 🆗:  Reference to docstring for ivy.function_name ([5.a](#ref1)) for the function description **and** modified `Parameters` and `Returns` sections as described in [the docs](https://unify.ai/docs/ivy/overview/deep_dive/docstrings.html#docstrings) in:
              - [ ] 🆗:  [ivy/array/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/general.py).
              - [ ] 🆗:  [ivy/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py) (in the static and instance method versions).
              - [ ] ⏩:   [ivy/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py) if the function has a special method  ( like `__function_name__` ).
              - [ ] ⏩:  [ivy/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py) if the function has a reverse special method  ( like `__rfunction_name__` ).
              - [ ] ⏩: [ivy/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py) if the function has a special method ( like `__function_name__` ).
              - [ ] ⏩:  [ivy/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py) if the function has a reverse special method  ( like `__rfunction_name__` ).
6. - [ ] 🆗: Add thorough  [Docstring Examples](https://unify.ai/docs/ivy/overview/deep_dive/docstring_examples.html#docstring-examples)  for every function  **and**  its  _relevant methods_  and ensure they pass the docstring tests.

		**Functional Examples** in [ivy/functional/general.py](https://github.com/unifyai/ivy/blob/main/ivy/functional/ivy/general.py).

		1. - [ ] 🆗: Cover all possible variants for each of the arguments independently (not combinatorily).
	 	2. - [ ] 🆗: Vary the values and input shapes considerably between examples.
	 	3. - [ ] 🆗: Start out simple and get more complex with each example.
	 	4. - [ ] 🆗: Show an example with:
			   - [ ] 🆗: `out` unused.
			   - [ ] 🆗: `out` used to update a new array y.
			   - [ ] 🆗: `out` used to inplace update the input array x (if x has the same dtype and shape as the return).
	 	5. - [ ] 🆗: If broadcasting is relevant for the function, then show examples which highlight this.

		**Nestable Function Examples** in [ivy/functional/general.py](https://github.com/unifyai/ivy/blob/main/ivy/functional/ivy/general.py).
		Only if the function supports nestable operations.

	 	6. - [ ] 🆗: <a name="ref2"></a> Add an example that passes in an  `ivy.Container`  instance in place of one of the arguments.
	 	7. - [ ] 🆗: <a name="ref3"></a> Add an example passes in  `ivy.Container`  instances for multiple arguments.

		**Container Static Method Examples** in [ivy/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py).

	 	8. - [ ] 🆗: The example from point ([6.f](#ref2)) should be replicated, but added to the  `ivy.Container`  **static method** docstring in with  `ivy.<func_name>`  replaced with  `ivy.Container.static_<func_name>`  in the example.
	 	9. - [ ] 🆗: The example from point ([6.g](#ref3)) should be replicated, but added to the  `ivy.Container`  **static method** docstring, with  `ivy.<func_name>`  replaced with  `ivy.Container.static_<func_name>`  in the example.

		**Array Instance Method Example** in [ivy/array/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/general).

		10. - [ ] 🆗: Call this instance method of the  `ivy.Array`  class.

		**Container Instance Method Example** in [ivy/container/general.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/general.py).

		11. - [ ] 🆗: Call this instance method of the  `ivy.Container`  class.

		**Array Operator Examples** in [ivy/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py).

		12. - [ ] ⏩: Call the operator on two  `ivy.Array`  instances.
	 	13. - [ ] ⏩: Call the operator with an  `ivy.Array`  instance on the left and  `ivy.Container`  on the right.

		**Array Reverse Operator Example** in [ivy/array/array.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/array/array.py).

		14.  - [ ] ⏩: Call the operator with a  `Number`  on the left and an  `ivy.Array`  instance on the right.

		**Container Operator Examples** in [ivy/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py).

		15. - [ ] ⏩: Call the operator on two `ivy.Container` instances containing Number instances at the leaves.
	 	16. - [ ] ⏩: Call the operator on two `ivy.Container` instances containing `ivy.Array` instances at the leaves.
	 	17. - [ ] ⏩: Call the operator with an `ivy.Container` instance on the left and `ivy.Array` on the right.

		**Container Reverse Operator Example** in [ivy/container/container.py](https://github.com/unifyai/ivy/blob/main/ivy/data_classes/container/container.py).

		18. - [ ] ⏩: Following example in the [`ivy.Container.__radd__`](https://github.com/unifyai/ivy/blob/e28a3cfd8a4527066d0d92d48a9e849c9f367a39/ivy/container/container.py#L173) docstring, with the operator called with a `Number` on the left and an `ivy.Container` instance on the right.

		**Tests**

		19. - [ ] ✅: Docstring examples tests passing.
		20. - [ ] ✅: Lint checks passing.

**Comment by Kacper-W-Kozdon** [comment]
Hi @AnnaTz ,
the issue #26498 was closed by you with my previous pull request but I noticed that ivy had an inplace_update method (which I'll have to check for what exactly it does)- but I'm guessing that it ensures compatibility of the output with other functions and methods from ivy, so I replaced the rudimentary and limited bit of code from the original PR with ivy's function. Nothing more was changed. And I attached fresh debug log from pytest this time- the only warnings come from the tensorflow backend, which was not changed.
[pytestdebug.log](https://github.com/unifyai/ivy/files/13452879/pytestdebug.log)

**Review by AnnaTz — APPROVED** [review]

**Comment by AnnaTz** [comment]
> Hi @AnnaTz , the issue #26498 was closed by you with my previous pull request but I noticed that ivy had an inplace_update method (which I'll have to check for what exactly it does)- but I'm guessing that it ensures compatibility of the output with other functions and methods from ivy, so I replaced the rudimentary and limited bit of code from the original PR with ivy's function. Nothing more was changed. And I attached fresh debug log from pytest this time- the only warnings come from the tensorflow backend, which was not changed. [pytestdebug.log](https://github.com/unifyai/ivy/files/13452879/pytestdebug.log)

Right, don't know how I missed that in your previous PR. Thanks for being thorough enough to correct this!

**Comment by Kacper-W-Kozdon** [comment]
@AnnaTz Ok, I found out something weird- for some reason my pytest was not running properly (it did not collect all backends for testing- if you check the pytestdebug.log, it has "collecting 1 items" and after i reinstalled and updated some libraries, it now has "collecting 5 items"), so there is still some problem with the implementation. I fixed one of the issues already, I have to figure out from the implementations in jax and numpy, where the problem is, because they also do not have the argument for "batch_dims" and thus require basically the same approach with some small modifications to syntax. I think I might've already found the problem already, but I'm still checking. Should I open a new issue or just a pull request and tag you in it? I'll attach pytestdebug.log for comparison.

**Comment by AnnaTz** [comment]
> @AnnaTz Ok, I found out something weird- for some reason my pytest was not running properly (it did not collect all backends for testing- if you check the pytestdebug.log, it has "collecting 1 items" and after i reinstalled and updated some libraries, it now has "collecting 5 items"), so there is still some problem with the implementation. I fixed one of the issues already, I have to figure out from the implementations in jax and numpy, where the problem is, because they also do not have the argument for "batch_dims" and thus require basically the same approach with some small modifications to syntax. I think I might've already found the problem already, but I'm still checking. Should I open a new issue or just a pull request and tag you in it? I'll attach pytestdebug.log for comparison. Sorry for the issue! I was not aware my pytest was set improperly.
> 
> [UPDATE] It's almost fixed, I still need to figure out the batch_dims == 0 case, because torch.gather() does not accept indices differing dimensionally from params and torch.take() prioritizes the shape of indices when shaping the output, so neither fully reconstructs tf.gather() results and need something to take care of shape.
> 
> [UPDATE2] I figured out that tensorflow's implementation is taken as the ground truth for tests. torch.gather() works well as long as the shapes match. Both the choice of axis and batch_dims should be solved by just doing indices.expand() and params.expand() according to the rules from tensorflow's documentation. So far, it seems to for batch_dims == 0. for batch_dims != 0 it will probably work as well, since all it does is looping through slices of params and indices. I should be able to finish tomorrow and with this the torch backend for gather should work just fine.

Hey @Kacper-W-Kozdon, feel free to open a new PR and tag me. Thanks for looking into this!

**Comment by Kacper-W-Kozdon** [comment]
One more update to the progress: torch.expand() on its own is problematic, because it doesn't like non-singleton dimensions being inserted inbetween non-singleton ones (so (2, 2) -> (2, 2, 1, 2) will likely throw an error) but it's easily solved by reshaping first with singleton dims (so (2, 2) -reshape-> (2, 1, 1, 2) -expand-> (2, 2, 1, 2)). This should be enough to deal with dimensionality problem of the native torch.gather() and make it mimic tensorflow's behaviour in ivy's torch backend.

**Comment by Kacper-W-Kozdon** [comment]
Update: it seems that the problems with dimensions, batch_dims and axis (for positive values of axis) have been solved while keeping the backend-specific implementation, now i'm removing the issues torch has with negative values of axis and if nothing else comes up, torch backend in ivy will be implemented with its native backend.

**Comment by Kacper-W-Kozdon** [comment]
Update: I think the implementation is working properly now, you can find it in: https://github.com/Kacper-W-Kozdon/notebook-testing-ivy/blob/main/Testing_Python.ipynb - I'll clean it from the comments and deal with the linting tomorrow, and then I'll also do a pull request (if I see any more failing tests, I'll just ask for some help at this point). But the implementation uses only torch this time. One important thing @AnnaTz - and it might be important for *all* of the tests using tensorflow as the ground truth- tensorflow's documentation seems to be imprecise and can omit some unique behaviours for edge cases. For example, for tf.gather(), if indices[batch_dims:].shape == (0), the formula from the documentation to get the shape of the result is *wrong*; for example, while the output of the tf.gather() has the shape (1, 1), the formula will give the shape equal to (1, ).
